### DISCUSSION TITLE
Viewing an anndata from a public bucket

### DISCUSSION THREAD

**Discussion by ivirshup** [vitessce/vitessce-python#365] [Q&A]
Hey! I am running into some trouble visualizing an anndata in Zarr format that I've uploaded to a public bucket. Here's what I'm running:

```python
import zarr
from vitessce import (
    VitessceWidget,
    VitessceConfig,
    Component as cm,
    CoordinationType as ct,
    AnnDataWrapper,
)
TEST_ANNDATA_URI = "s3://tmp-census-zarr/v1/homo_sapiens/0041b9c3-6a49-4bf7-8514-9bc7190067a7.zarr"

vc = VitessceConfig(schema_version="1.0.15")
dataset = vc.add_dataset().add_object(
        AnnDataWrapper(
                adata_store=z.store,
                obs_embedding_paths=["obsm/X_umap"],
                obs_embedding_names=["UMAP"],
                obs_set_paths=["obs/cell_type"],
                obs_set_names=["cell_type"],
                obs_feature_matrix_path="X",
        )
)
scatterplot = vc.add_view(cm.SCATTERPLOT, dataset=dataset, mapping="UMAP")
cell_sets = vc.add_view(cm.OBS_SETS, dataset=dataset)
genes = vc.add_view(cm.FEATURE_LIST, dataset=dataset)
heatmap = vc.add_view(cm.HEATMAP, dataset=dataset)
layout = vc.layout((scatterplot | (cell_sets / genes)) / heatmap)

vc.web_app()
```

But the webapp is blank. AFAICT the web app is trying to retrieve the data from `localhost` even though I've passed a zarr store backed by s3. Here are the console errors that make me think that:

![image](https://github.com/user-attachments/assets/ebac5954-242b-40c2-b33e-178fbdfaa454)

I've cobbled together the code above from a combination of the [s3 data export tutorial](https://python-docs.vitessce.io/notebooks/data_export_s3.html#6.-View-on-vitessce.io) and the [zarr tutorial](https://python-docs.vitessce.io/data_options.html#jupyter-process-anywhere-files-anywhere-that-can-be-accessed-via-zarr).

What am I supposed to be doing differently here?

Thanks!

**Comment by keller-mark** [comment]
Have you tried explicitly using FSStore? It seems to work via the HTTP URL but throws an error about S3 credentials for the `s3://` protocol store.

Another note is that passing a zarr store via `adata_store` is not compatible with `vc.web_app()` - only with `vc.widget()`. This is because the zarr data is passed using a widget-based RPC mechanism.

```python
import zarr
from vitessce import (
    VitessceWidget,
    VitessceConfig,
    Component as cm,
    CoordinationType as ct,
    AnnDataWrapper,
)
s3_store = zarr.storage.FSStore("s3://tmp-census-zarr/v1/homo_sapiens/0041b9c3-6a49-4bf7-8514-9bc7190067a7.zarr")
http_store = zarr.storage.FSStore("https://tmp-census-zarr.s3.amazonaws.com/v1/homo_sapiens/0041b9c3-6a49-4bf7-8514-9bc7190067a7.zarr")

vc = VitessceConfig(schema_version="1.0.15")
dataset = vc.add_dataset().add_object(
        AnnDataWrapper(
                adata_store=http_store,
                obs_embedding_paths=["obsm/X_pca"],
                obs_embedding_names=["PCA"],
                obs_set_paths=["obs/cell_type"],
                obs_set_names=["cell_type"],
                obs_feature_matrix_path="X",
                feature_labels_path="var/feature_name",
                coordination_values={
                    "featureLabelsType": "gene_symbol"
                }
        )
)
scatterplot = vc.add_view(cm.SCATTERPLOT, dataset=dataset, mapping="UMAP")
cell_sets = vc.add_view(cm.OBS_SETS, dataset=dataset)
genes = vc.add_view(cm.FEATURE_LIST, dataset=dataset)
#heatmap = vc.add_view(cm.HEATMAP, dataset=dataset)
layout = vc.layout((scatterplot | (cell_sets / genes)))

vw = vc.widget()
vw
```

**Reply by ivirshup** [reply]
Thanks! I think this is the most important thing to me:

> Another note is that passing a zarr store via adata_store is not compatible with vc.web_app() - only with vc.widget().

I would like to create static web pages that let people look at the anndatas in a public s3 store. I was hoping to do this via the Python API.  Is this possible?

I saw I was able to manipulate the URL generated by `vc.web_app()` to make this work, but I has to replace the data url it had.

**Reply by keller-mark** [reply]
If the Zarr store is publicly accessible via HTTP on S3, then you should be able to use `adata_url` (instead of `adata_store`):

```python
vc = VitessceConfig(schema_version="1.0.15")
dataset = vc.add_dataset().add_object(
        AnnDataWrapper(
                adata_url="https://tmp-census-zarr.s3.amazonaws.com/v1/homo_sapiens/0041b9c3-6a49-4bf7-8514-9bc7190067a7.zarr",
                obs_embedding_paths=["obsm/X_pca"],
                obs_embedding_names=["PCA"],
                obs_set_paths=["obs/cell_type"],
                obs_set_names=["cell_type"],
                obs_feature_matrix_path="X",
                feature_labels_path="var/feature_name",
                coordination_values={
                    "featureLabelsType": "gene_symbol"
                }
        )
)
scatterplot = vc.add_view(cm.SCATTERPLOT, dataset=dataset, mapping="UMAP")
cell_sets = vc.add_view(cm.OBS_SETS, dataset=dataset)
genes = vc.add_view(cm.FEATURE_LIST, dataset=dataset)
#heatmap = vc.add_view(cm.HEATMAP, dataset=dataset)
layout = vc.layout((scatterplot | (cell_sets / genes)))

vw = vc.widget()
vw
```

For background, `adata_store` is intended for `vc.widget()` situations in which the Zarr store is not publicly accessible via HTTP, but a Zarr store instance can be constructed some other way, such as on an HPC cluster.
